### PR TITLE
Housing company state refactor

### PR DESCRIPTION
**Pull request by matti-lamppu**
# Hitas Pull Request

# Description

Replaces Housing company state with four fields: `regulation_status`, `exclude_from_statistics`, `over_thirty_years_old`, and `completed`. The first two are editable fields, while the last two are computable fields. Housing company still retains the state field until frontend changes are made.

## Pull request checklist

Check the boxes for each DoD item that has been completed:

- **Testing**
  - [x] Changes have been tested
  - [x] Automatic tests have been added
- **Database**
  - [x] Database migrations will work in the DEV & TEST environments
  - [x] initial.json has been updated to work with migrations
  - [x] Oracle migration has been updated
- **Documentation**
  - [ ] Tooltips have been added in the frontend for all new fields
  - [x] OpenAPI definitions have been updated
  - [ ] Test instructions have been written for the customer in the appropriate ticket in Jira
  - [ ] Terminology page in Confluence has been updated

## Test plan

- [x] Automated tests
- [x] Run migration without issues
- [x] All old Housing Company `state`-related stuff works with the new fields

## Tickets

This pull request resolves all or part of the following ticket(s): HT-510

### CODE REVIEW

File: backend/hitas/models/housing_company.py [hunk 1]
```diff
@@ -95,6 +114,29 @@ def total_shares_count(self) -> int:
...
+        from hitas.models import Apartment
+
+        completion_dates: list[datetime.date | None] = Apartment.objects.filter(
+            building__real_estate__housing_company=self
+        ).values_list("completion_date", flat=True)
+
+        if None in completion_dates:
+            return False
+
+        return relativedelta(timezone.now().date(), max(completion_dates)).years >= 30
```
Comment (ranta) [inline backend/hitas/models/housing_company.py:128]: I would change the property name to signal the the return value is a boolean. Also, there shouldn't be a need to do the checking in python, as we can just ask the DB to return the needed value like so:

```py
    @property
    def is_over_thirty_years_old(self) -> bool:
        from hitas.models import Apartment

        completion_date: Optional[datetime.date] = (
            Apartment.objects.filter(building__real_estate__housing_company=self)
            .order_by("-completion_date")
            .first()
            .completion_date
        )

        if completion_date is None:
            return False

        return relativedelta(timezone.now().date(), completion_date).years >= 30
```

File: backend/hitas/models/housing_company.py [hunk 2]
```diff
@@ -95,6 +114,29 @@ def total_shares_count(self) -> int:
...
+
+    @property
+    def completed(self) -> bool:
+        from hitas.models import Apartment
+
+        completion_dates: list[datetime.date | None] = Apartment.objects.filter(
+            building__real_estate__housing_company=self
+        ).values_list("completion_date", flat=True)
+
+        return None not in completion_dates
```
Comment (ranta) [inline backend/hitas/models/housing_company.py:138]: Here the same thing as above, we don't need to do the checking in python, we can simply:
```py
def is_completed(self) -> bool:
    completion_date = ...
    return bool(completion_date)
```

File: backend/hitas/tests/factories/housing_company.py [hunk 3]
```diff
@@ -22,6 +23,8 @@ class Meta:
     display_name = factory.Sequence(lambda n: f"Test Housing company {n:03}")
     official_name = factory.LazyAttribute(lambda self: f"As Oy {self.display_name}")
     state = fuzzy.FuzzyChoice(state[0] for state in HousingCompanyState.choices())
+    exclude_from_statistics = False
+    regulation_status = fuzzy.FuzzyChoice(state[0] for state in RegulationStatus.choices())
```
Comment (ranta) [inline backend/hitas/tests/factories/housing_company.py:27]: I'd rather have the `regulation_status` default value be `RegulationStatus.REGULATED`, so it doesn't need to be specified for every case separately.
Then when we need a non-regulated housing company we can specify it.

File: backend/hitas/oracle_migration/utils.py [hunk 4]
```diff
@@ -80,6 +81,18 @@ def housing_company_state_from(code: str) -> HousingCompanyState:
...
+def housing_company_regulation_status_from(code: str) -> RegulationStatus:
+    # These are hardcoded as the code number (C_KOODISTOID) and
+    # the name (C_NAME) are the only ones that can be used to
+    # identify these types
+    if code == "003":
+        return RegulationStatus.RELEASED_BY_HITAS
+    if code == "004":
+        return RegulationStatus.RELEASED_BY_PLOT_DEPARTMENT
+
+    return RegulationStatus.REGULATED
```
Comment (ranta) [inline backend/hitas/oracle_migration/utils.py:93]: Should `005` (`Half-Hitas`) be regulated or not? I'm not 100% sure which one it should be 🤔

File: backend/hitas/views/housing_company.py [hunk 5]
```diff
@@ -45,6 +49,18 @@
 from hitas.views.utils.serializers import YearMonthSerializer
 
 
+class RegulationStatusViewSet(ViewSet):
+    def list(self, request: Request, *args, **kwargs) -> Response:
+        data = [
+            {
+                "name": regulation_status.value,
```
Comment (ranta) [inline backend/hitas/views/housing_company.py:56]: `value` or `name`?

`name` sounds kind of synonymous with `label` to me 🤔

File: backend/hitas/oracle_migration/utils.py (hunk 4 above)
Comment (matti-lamppu) [reply backend/hitas/oracle_migration/utils.py:93]: If I remember correctly, half-Hitas housing companies are only regulated for the initial sale, and not afterward.

File: backend/hitas/tests/factories/housing_company.py (hunk 3 above)
Comment (matti-lamppu) [reply backend/hitas/tests/factories/housing_company.py:27]: Sure, I'll change it

File: backend/hitas/models/housing_company.py (hunk 1 above)
Comment (matti-lamppu) [reply backend/hitas/models/housing_company.py:128]: Yes, I'll change this.

File: backend/hitas/models/housing_company.py (hunk 2 above)
Comment (matti-lamppu) [reply backend/hitas/models/housing_company.py:138]: Changed.

File: backend/hitas/views/housing_company.py (hunk 5 above)
Comment (matti-lamppu) [reply backend/hitas/views/housing_company.py:56]: I guess we can go with the value-label separation enums use. I'll change it.

**Review by ranta — APPROVED** [review]
🐴

**Pull request merged**